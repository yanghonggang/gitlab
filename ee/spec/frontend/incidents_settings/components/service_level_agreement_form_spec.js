import { mount, shallowMount } from '@vue/test-utils';
import { GlButton, GlForm, GlFormGroup } from '@gitlab/ui';
import ServiceLevelAgreementForm from 'ee/incidents_settings/components/service_level_agreement_form.vue';
import merge from 'lodash/merge';

const defaultData = { enabled: true, duration: 15, units: 'minutes' };

describe('Alert integration settings form', () => {
  let wrapper;
  const service = { updateSettings: jest.fn().mockResolvedValue() };

  const mountComponent = (options, mountMethod = shallowMount) => {
    wrapper = mountMethod(
      ServiceLevelAgreementForm,
      merge(
        {
          provide: {
            service,
            serviceLevelAgreementSettings: { available: true },
          },
          data() {
            return { ...defaultData };
          },
        },
        options,
      ),
    );
  };

  afterEach(() => {
    if (wrapper) {
      wrapper.destroy();
      wrapper = null;
    }
  });

  const findForm = () => wrapper.find(GlForm);
  const findFormGroup = () => wrapper.find(GlFormGroup);
  const findSubmitButton = () => wrapper.find(GlButton);

  it('renders an empty component when feature not available', () => {
    mountComponent({ provide: { serviceLevelAgreementSettings: { available: false } } });

    expect(wrapper.html()).toBe('');
  });

  it('should match the default snapshot', () => {
    mountComponent();

    expect(wrapper.element).toMatchSnapshot();
  });

  describe('Form fields', () => {
    it('should call service `updateSettings` on submit', async () => {
      mountComponent({}, mount);

      findForm().trigger('submit');

      expect(service.updateSettings).toHaveBeenCalledWith(
        expect.objectContaining({
          sla_timer: true,
          sla_timer_minutes: 15,
        }),
      );
    });

    it('should allow submit when form is valid', () => {
      mountComponent();

      expect(findFormGroup().attributes('state')).toBe('true');
    });

    describe.each`
      enabled | duration | unit         | isValid  | expectedMessage
      ${true} | ${''}    | ${'minutes'} | ${false} | ${'Time limit must be a valid number'}
      ${true} | ${-15}   | ${'minutes'} | ${false} | ${'Time limit must be greater than 0'}
      ${true} | ${0}     | ${'minutes'} | ${false} | ${'Time limit must be greater than 0'}
      ${true} | ${0.5}   | ${'minutes'} | ${false} | ${'Time limit must be a multiple of 15 minutes'}
      ${true} | ${5}     | ${'minutes'} | ${false} | ${'Time limit must be a multiple of 15 minutes'}
      ${true} | ${15}    | ${'minutes'} | ${true}  | ${''}
      ${true} | ${0}     | ${'hours'}   | ${false} | ${'Time limit must be greater than 0'}
      ${true} | ${0.15}  | ${'hours'}   | ${false} | ${'Time limit must be a multiple of 15 minutes'}
      ${true} | ${0.5}   | ${'hours'}   | ${true}  | ${''}
      ${true} | ${1}     | ${'hours'}   | ${true}  | ${''}
      ${true} | ${24}    | ${'hours'}   | ${true}  | ${''}
    `(
      'Inputs enabled "$enabled", "$duration" $unit',
      ({ enabled, duration, unit, isValid, expectedMessage }) => {
        beforeEach(() => {
          mountComponent(
            {
              data() {
                return { ...defaultData, enabled, duration, unit };
              },
            },
            mount,
          );
        });

        it(`should${isValid ? '' : ' not'} allow submit`, () => {
          expect(findSubmitButton().attributes('disabled')).toBe(isValid ? undefined : 'disabled');
        });

        it(`should show ${isValid ? 'no' : 'the correct'} error message`, () => {
          expect(findFormGroup().text()).toContain(expectedMessage);
        });
      },
    );
  });
});
