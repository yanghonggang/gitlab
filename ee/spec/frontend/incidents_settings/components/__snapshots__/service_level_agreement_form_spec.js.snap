// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert integration settings form should match the default snapshot 1`] = `
<gl-tab-stub
  title="Incident settings"
  titlelinkclass=""
>
  <gl-form-stub
    class="gl-pt-3"
  >
    <p
      class="gl-line-height-20"
    >
      
      You may choose to introduce a countdown timer in incident issues to better track Service Level Agreements (SLAs). The timer is automatically started when the incident is created, and sets a time limit for the incident to be resolved in. When activated, "time to SLA" countdown will appear on all new incidents.
    
    </p>
     
    <gl-form-checkbox-stub
      checked="true"
      class="gl-my-4"
    >
      <span>
        Activate "time to SLA" countdown timer
      </span>
       
      <gl-form-text-stub
        class="gl-font-base gl-text-gray-400"
        tag="small"
        textvariant="muted"
      >
        
        When activated, this will apply to all new incidents within the project
      
      </gl-form-text-stub>
    </gl-form-checkbox-stub>
     
    <gl-form-group-stub
      invalid-feedback=""
      label="Time limit"
      label-for="sla-duration"
      state="true"
    >
      <div
        class="gl-display-flex gl-flex-direction-row"
      >
        <gl-form-input-stub
          id="sla-duration"
          number=""
          size="xs"
          step="15"
          type="number"
          value="15"
        />
         
        <gl-form-select-stub
          class="gl-w-auto gl-ml-3 gl-line-height-normal gl-border-gray-400"
          options="[object Object],[object Object]"
          value="minutes"
        />
      </div>
       
      <gl-form-text-stub
        class="gl-font-base gl-text-gray-400!"
        tag="small"
        textvariant="muted"
      >
        
          Time limit must be a multiple of 15 minutes
        
      </gl-form-text-stub>
    </gl-form-group-stub>
     
    <gl-button-stub
      buttontextclasses=""
      category="primary"
      icon=""
      size="medium"
      type="submit"
      variant="success"
    >
      
      Save changes
    
    </gl-button-stub>
  </gl-form-stub>
</gl-tab-stub>
`;
